### DISCUSSION TITLE
How to call custom middleware in this example

### DISCUSSION THREAD

**Discussion by kennyth01** [honojs/hono#2440] [General]
Hi all,
I have this script that calls sessionMiddleware:

```
import { sessionMiddleware, CookieStore, Session } from 'hono-sessions'

const app = new Hono<{ Bindings: Env, Variables: {
    session: Session,
    session_key_rotation: boolean
} }>()
const store = new CookieStore()

app.use('*', 
sessionMiddleware({
    store,
    encryptionKey: 'some_string_thritytwo_characters',
    expireAfterSeconds: 900,
    cookieOptions: {
      sameSite: 'Lax',
      path: '/',
      httpOnly: true,
    },
  }),
    async (c, next) => {
    console.log(`inside all [${c.req.method}] ${c.req.url}`)

    await next()
  })
```

This is working perfectly!

However, I want it to call inside the method like this so I can get the Context object

```
app.use('*',  async (c, next) => {
    console.log(`inside all [${c.req.method}] ${c.req.url}`)
sessionMiddleware({
    store,
    encryptionKey: c.env.MY_KEY,
    expireAfterSeconds: 900,
    cookieOptions: {
      sameSite: 'Lax',
      path: '/',
      httpOnly: true,
    },
  })
    await next()
  })
```
What is the correct way? The second snippet is not working correctly.

**Comment by NicoPlyley** [comment]
Are you looking to get access to the context inside your custom middleware?

Is so the context is passed to custom middleware but it does not have type safety. You can use the [createMiddleware](https://hono.dev/helpers/factory#createmiddleware) wrapper function which will add types to the context.

**Comment by kennyth01** [comment]
hi @NicoPlyley , Unfortunately, I'm not using custom middleware, I am using a 3rd party middleware called `sessionMiddleware ` from hono-sessions

Yes, I want to get the context.

**Reply by NicoPlyley** [reply]
What are you trying to achieve with the context? Maybe I can help you find a work around.

Is it to get the variable for secret key?

**Reply by kennyth01** [reply]
yes correct, i want to get the env variable. my env is inside Context object

**Reply by NicoPlyley** [reply]
Where it is coming from in the first place, the only way I see is moving it to another location

**Reply by kennyth01** [reply]
What do u mean where is it coming from? Are you referring to this: `c.env.MY_KEY`

If yes, it'll be something like:

```
export type Env = {
    DATABASE_URL: string
    MY_KEY: string
}
const app = new Hono<{ Bindings: Env, Variables: {
    session: Session,
    session_key_rotation: boolean
} }>()
const store = new CookieStore()

app.use('*',  async (c, next) => {
    console.log(`inside all [${c.req.method}] ${c.req.url}`)
sessionMiddleware({
    store,
    encryptionKey: c.env.MY_KEY,
    expireAfterSeconds: 900,
    cookieOptions: {
      sameSite: 'Lax',
      path: '/',
      httpOnly: true,
    },
  })
    await next()
  })
```

**Reply by NicoPlyley** [reply]
Are you using CloudFlare Workers where you can only access env var in the context? If so, I would open an issue for hono-sessions to allow c.env names, in a parameter like `encryptionKeyFromCtx`. If you want a hacky way to make it work. Create a empty let out side any function to scope it globally. Make a middleware that saves the ctx env to that let. Then use that variable in the next middleware for the session

Very unconventional, but it should work. Ultimately this would be an issue on hono-sessions side.

**Reply by kennyth01** [reply]
thanks @NicoPlyley ,

I've manage to make it work via the following way:

```
export const cookieSessionMiddleware = async (c: Context, next: undefined) => {
    const store = new CookieStore()

    const m = sessionMiddleware({
        store,
        encryptionKey: c.env.SESSION_ENCRYPTION_KEY,
        expireAfterSeconds: 900,
        cookieOptions: {
          sameSite: 'Lax',
          path: '/',
          httpOnly: true,
        },
      })

      return m(c, next)
  }
```

then use this middleware:
```
app
.use('*')
.use(cookieSessionMiddleware)
```

**Reply by NicoPlyley** [reply]
Well glad that works you can use the createMiddleware to gives types to the context and next
